include/jemalloc/internal/witness_inlines.h

```diff
@@ -92,14 +96,28 @@ witness_assert_lockless(tsdn_t *tsdn) {
 	}
 	tsd = tsdn_tsd(tsdn);
 
+	d = 0;
 	witnesses = tsd_witnessesp_get(tsd);
 	w = ql_last(witnesses, link);
 	if (w != NULL) {
-		witness_lockless_error(witnesses);
+		ql_reverse_foreach(w, witnesses, link) {
+			if (w->rank < rank_inclusive) {
+				break;
+			}
+			d++;
```

Comment:
In practice I haven't found this to be a common problem, and in fact now that I think about it the error-prone cases all involve ring mutation during iteration.  The foreach macros basically can't fail unless the ring is outright corrupt (links traverse into a cycle that does not include the element of origin).

That said, I'm open to adding sanity checks.  A good one would be assuring that linkage is initialized prior to insertion; I do continue to get burned by forgetting to call ```qr_new()```/```ql_elm_new()```.